File: t17.py
# -*- conding:utf-8 -*-
#Author:lyc
import os, sys

# 题目：
# 给定一个仅包含数字 2-9 的字符串，返回所有它能表示的字母组合。
# 给出数字到字母的映射如下（与电话按键相同）。注意 1 不对应任何字母。
#
# 示例:
# 输入："23"
# 输出：["ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"].
# 说明:
# 尽管上面的答案是按字典序排列的，但是你可以任意选择答案输出的顺序。

# 递归解法
class Solution:
    def letterCombinations(self, digits: str) -> [str]:

        str_dic = {
            '2': 'abc',
            '3': 'def',
            '4': 'ghi',
            '5': 'jkl',
            '6': 'mno',
            '7': 'pqrs',
            '8': 'tuv',
            '9': 'wxyz'
        }

        result = []
        if not digits:
            return []
        for i in str_dic[digits[0]]:
            if len(digits) == 1:
                result.append(i)
                if i == str_dic[digits[0]][-1]:
                    return result
            else:
                for j in self.letterCombinations(digits[1:]):
                    result.append(i+j)
                    if i == str_dic[digits[0]][-1] and j == self.letterCombinations(digits[1:])[-1]:
                        return result


# 快速递归法
class Solution:
    def letterCombinations(self, digits: str) -> [str]:

        str_dic = {
            '2': 'abc',
            '3': 'def',
            '4': 'ghi',
            '5': 'jkl',
            '6': 'mno',
            '7': 'pqrs',
            '8': 'tuv',
            '9': 'wxyz'
        }

        result = []
        if not digits:
            return result
        elif len(digits) == 1:
            return list(str_dic[digits[0]])
        else:
            res = self.letterCombinations(digits[:-1])
            for i in res:
                for j in str_dic[digits[-1]]:
                    result.append(i+j)

        return result


result = Solution()
x = '234'
print(result.letterCombinations(x))

File: t17.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <string>
#include <map>

using namespace std;

/**
 *  leetcode第17题
 *  递归
 *  Author:lyc
*/

// 解法一：递归
class Solution {
public:
    vector<string> letterCombinations(string digits) 
    {

        // 定义电话键盘对应map
        map<char,vector<string>> str_dic = {
                {'2',{"a","b","c"}},
                {'3',{"d","e","f"}},
                {'4',{"g","h","i"}},
                {'5',{"j","k","l"}},
                {'6',{"m","n","o"}},
                {'7',{"p","q","r","s"}},
                {'8',{"t","u","v"}},
                {'9',{"w","x","y","z"}}
        };
        
        // 声明结果vector
        vector<string> result;

        // 判断输入字符为空时直接返回空vector
        if(digits.size() < 1){
            return result;
        };

        // 输入字符串长度
        int digOflens = digits.size();

        for(auto s:str_dic[digits[0]])
        {
            // 当输入字符只有一个时，逐个添加至result并返回
            if(digits.size() == 1)
            {
                result.push_back(s);
                if(s == str_dic[digits[0]].back())
                {
                    return result;
                }
            }else{
                // 截取输入字符串并递归调用
                string tail = digits.substr(1,digits.size());
                for(auto n:letterCombinations(tail))
                {
                    result.push_back(s+n);
                    // 判断输入字符串到最后一位及对应的str_dic中也为最后一个时函数返回
                    if(s == str_dic[digits[0]].back() & n == letterCombinations(tail).back())
                    {
                        return result;
                    };
                };
            };
        };
        // 函数必须有返回值
        return result;
    };
};


// 解法二：快速递归
class Solution {
public:
    vector<string> letterCombinations(string digits) 
    {

        // 定义电话键盘对应map
        map<char,vector<string>> str_dic = {
                {'2',{"a","b","c"}},
                {'3',{"d","e","f"}},
                {'4',{"g","h","i"}},
                {'5',{"j","k","l"}},
                {'6',{"m","n","o"}},
                {'7',{"p","q","r","s"}},
                {'8',{"t","u","v"}},
                {'9',{"w","x","y","z"}}
        };
        
        // 声明结果vector
        vector<string> result;

        // 判断输入字符为空时直接返回空vector
        if(digits.size() < 1)
        {
            return result;
        }
        else if(digits.size() == 1)
        {
            return str_dic[digits[0]];
        }
        else
        {
            vector<string> res = letterCombinations(digits.substr(0,digits.size()-1));
            for(auto i:res)
            {
                for(auto j:str_dic[digits.back()])
                {
                    result.push_back(i+j);
                }
            }
        }
        return result;
    };
};




int main()
{
    Solution model;

    string x = "234";

    vector<string> res = model.letterCombinations(x);

    // 打印结果
    for(auto i:res){
        cout << i << ',';
    };
    cout << endl;

    system("pause");
    return 0;
}